v0/destinations/google_ads/transform.js

```diff
@@ -0,0 +1,169 @@
+ *
+ * @param {Rudder Payload} message
+ * To set app_event_name
+ * The name of any custom app event which is not accepted in the app_event_type field.
+ * This field should contain 1 to 64 Unicode characters (using UTF-8 encoding). This field is required if app_event_type is custom.
+ * https://developers.google.com/app-conversion-tracking/api/request-response-specs#conversion_tracking_request
+ *
+ */
+
+function setEventName(message) {
+  const defaultEventTypes = [
+    "first_open",
+    "session_start",
+    "in_app_purchase",
+    "view_item_list",
+    "view_item",
+    "view_search_results",
+    "add_to_cart",
+    "ecommerce_purchase",
+    "custom"
+  ];
+  const eventName = get(message, "event");
+  /**  check if the event name is one of the default event types.
+   * This field must not contain any of the values reserved for app_event_type.
+   * If a reserved event name is used, the API will return an APP_EVENT_NAME_RESERVED_VALUE error.
+   */
+  const isDefaultName = defaultEventTypes.indexOf(eventName) >= 0;
+  if (isDefaultName) {
+    throw new Error(
+      "Event name is a reserved event name in Google Ads. Aborting Message"
+    );
+  }
+  return eventName;
+}
+/**
+ *
+ * @param {Rudder Payload} message
+ * To set app_event_type
+ * The name of the app event that occurred. This field is an enumeration, and will only accept valued predefined by google ads.
+ * https://developers.google.com/app-conversion-tracking/api/request-response-specs#conversion_tracking_request
+ *
+ */
+function setEventType(message) {
+  const eventName = get(message, "event");
+  switch (eventName.toLowerCase()) {
+    case "application installed":
+      return "first_open";
+    case "application opened":
+      return "session_start";
+    case "order completed":
+      // Use in_app_purchase for purchases made through the native app store; use ecommerce_purchase for all other purchases.
+      if (
+        get(message, "properties.affiliation") === "App Store" ||
+        get(message, "properties.affiliation") === "Google Store"
+      ) {
+        return "in_app_purchase";
+      }
+      return "ecommerce_purchase";
+    case "product list viewed":
+      return "view_item_list";
+    case "product added":
+      return "add_to_cart";
+    case "poducts searched":
+      return "view_search_results";
+    case "product viewed":
+      return "view_item";
+    default:
+      return "custom";
+  }
+}
+function responseBuilderSimple(message, category, destination) {
+  const { Config } = destination;
+  const { devToken, linkId, version, lat, fallbackIDFAZero } = Config;
+  const payload = constructPayload(message, MAPPING_CONFIG[category.name]);
+  if (payload) {
+    payload.dev_token = devToken;
+    payload.link_id = linkId;
+    payload.app_event_type = setEventType(message);
+    // This field is required if app_event_type is custom.
+    if (payload.app_event_type === "custom") {
+      payload.app_event_name = setEventName(message);
```

Comment:
We're throwing `Error` from this method. We should do it as soon as possible. The `constructPayload` call is expensive, and we should make sure that we are making this call if everything is correct.

Reply:
done